Sources/HummingbirdCore/Server/HTTP/HTTPChannelHandler.swift

```diff
@@ -33,83 +33,77 @@ enum HTTPChannelError: Error {
+
+                        // read first part, verify it is a head
+                        guard let part = try await iterator.next() else { return }
+                        guard case .head(var head) = part else {
+                            throw HTTPChannelError.unexpectedHTTPPart(part)
                         }
-                        // set to idle unless it is cancelled then exit
-                        guard processingRequest.exchange(.idle, ordering: .relaxed) == .processing else { break }
 
-                        // Flush current request
-                        // read until we don't have a body part
-                        var part: HTTPRequestPart?
                         while true {
-                            part = try await iterator.next()
-                            guard case .body = part else { break }
+                            let bodyStream = NIOAsyncChannelRequestBody(iterator: iterator)
+                            let request = HBRequest(head: head, body: .init(asyncSequence: bodyStream))
+                            let response: HBResponse
+                            do {
+                                response = try await self.responder(request, asyncChannel.channel)
+                            } catch {
+                                response = self.getErrorResponse(from: error, allocator: asyncChannel.channel.allocator)
+                            }
+                            do {
+                                try await outbound.write(.head(response.head))
+                                let tailHeaders = try await response.body.write(responseWriter)
+                                try await outbound.write(.end(tailHeaders))
+                            } catch {
+                                throw error
+                            }
+                            if request.headers[.connection] == "close" {
+                                throw HTTPChannelError.closeConnection
+                            }
+
+                            // Flush current request
+                            // read until we don't have a body part
+                            var part: HTTPRequestPart?
+                            while true {
+                                part = try await iterator.next()
+                                guard case .body = part else { break }
+                            }
+                            // if we have an end then read the next part
+                            if case .end = part {
+                                part = try await iterator.next()
+                            }
+
+                            // if part is nil break out of loop
+                            guard let part else {
+                                break
+                            }
+
+                            // Process next request unless it is cancelled then exit
+                            if Task.isShuttingDownGracefully || Task.isCancelled {
```

Comment:
Cancellation check is moved down. I can imagine that _some_ clients might fail if we do respond, but don't consume the body before closing the connection.